implement a dialog for importing externally generated tracks from csv

AnniekStok opened the pull request:
Goals for this PR:

- [x] Import segmentations from zarr and tiff files
- [x] Automatically relabel segmentations if not unique across time
- [x] Import tracks from CSV with custom column names for required attributes (don't require exact string match to our internal names)
    - [x] Allow specifying a seg_id column if the segmentations are not unique across time
- [x] Have the option to compute area from the segmentation (if there is one) or use a column of the CSV, or don't have it at all
- [x] Make the user specify metadata:
    - [x] Dimensionality of data
    - [x] Existance of segmentation
    - [x] Scale of segmentation
- [x] Export custom features to csv from solution tracks

cmalinmayor commented on src/motile_plugin/import_export/load_tracks.py:
```diff
@@ -0,0 +1,96 @@
+            corresponds to the label ids in the segmentation array
+
+    Returns:
+        Tracks: a tracks object ready to be visualized with
```
I think we should delete view_external_tracks and just add_tracks to the tracks list

AnniekStok replied:
Done! I also updated the documentation and sample script.

cmalinmayor commented on src/motile_plugin/import_export/load_tracks.py:
```diff
@@ -0,0 +1,96 @@
+                attrs["seg_id"] = int(row[selected_columns["seg_id (if seg provided)"]])
+            for key in extra_columns:
+                if extra_columns[key] != "Select Column":
+                    attrs[key] = row[extra_columns[key]]
```
We will need to do something to determine the data type of the extra columns. I'm open to loading the CSV in pandas and letting it infer the types, if you think that makes sense. (Including replacing the `for row in reader` loop with a loop over a dataframe). Again, we should think about making attributes more generic and dealing with them in a consistent manner (and using pandas type inference is a step in that direction).

AnniekStok replied:
Switched to using pandas, which seems to indeed work fine for inferring the data types.

cmalinmayor commented on src/motile_plugin/import_export/load_tracks.py:
```diff
@@ -0,0 +1,96 @@
+    if segmentation is not None:
+        segmentation = np.expand_dims(
+            segmentation, axis=1
+        )  # do we need to keep the empty hypothesis dim?
```
Yes, for now! Until we update to motile_toolbox v0.4.0

cmalinmayor reviewed (changes requested):
I think this is mostly stuff we discussed Monday.  I updated the file structure a bit to make an import_export folder, and I didn't see any circular import issues when I ran it, but let me know if there is anything weird. I'll keep thinking about how we want to generally handle attributes in a more consistent way too.

AnniekStok commented on src/motile_plugin/data_views/views_coordinator/tracks_list.py:
```diff
@@ -94,11 +96,26 @@ def __init__(self):
         load_button = QPushButton("Load tracks")
         load_button.clicked.connect(self.load_tracks)
 
+        load_external_button = QPushButton("Import external tracks")
```
done!

AnniekStok commented on src/motile_plugin/import_export/export_solution_to_csv.py:
```diff
@@ -0,0 +1,57 @@
+            track_id = solution.get_track_id(node_id)
+            time = solution.get_time(node_id)
+            position = solution.get_position(node_id)
+            custom_attr = [
```
Fixed with required=False and empty string if None. If you load a csv with missing values, it will have 'nan' for those. We currently plot 'nan' as 0 for the area attribute.

AnniekStok commented on src/motile_plugin/import_export/import_external_tracks_dialog.py:
```diff
@@ -0,0 +1,312 @@
+        self.name_widget = QLineEdit(self.name)
+        name_layout.addWidget(self.name_widget)
+
+        # Scale layout
```
Moved the scaling to its own widget class

AnniekStok commented on src/motile_plugin/import_export/import_external_tracks_dialog.py:
```diff
@@ -0,0 +1,312 @@
+                df = pd.read_csv(csv_file, nrows=0)
+                self.csv_headers = list(df.columns)
+
+                # Populate dropdowns with CSV headers
```
Done!

AnniekStok commented:
@cmalinmayor  Some updates here: I have changed the import code to work with the unique node ids while still allowing importing from a seg that is not labeled by id but by track_id or something else. I also included tests. For export, I switched to the export_tracks function on SolutionTracks. I also made some changes in preparation of having multiple feature measurements (I have this working on a separate branch but I changed it here too even though we only have one measurement (area)).

cmalinmayor commented:
@AnniekStok A couple quick notes in addition to my pushed changes (casting the track id to int was the easiest way to fix my numpy float issue):

- Can we have an option to scale the points as well? In case people have coordinates in pixel space, not calibrated units. It can be a check box on the scaled units page that says "Scale the point coordiantes"
- Currently if you try and load a segmentation zarr from the browse pop up dialog, it won't let you actually select it, because it is a directory. I think it's possible in the browse dialog to allow selecting directories?
- It would be nice to add a tiny bit of text above the segmentation text box to say what kinds of segmentations are supported (.tif, .tiff, .zarr)
- The option to compute area comes up even if you dont load a segmentation. I'm okay with this if it makes sense when we have more features, but it's a little wonky now

AnniekStok commented:
Some updates:

- points are now scaling automatically according to the x scale value. I also re-enabled setting it in the GUI, it seemed to be fine and I did not encounter the off-target effect any more (have to test more thoroughly still). 
-  I changed the segmentation browse menu so that you can either choose a tiff file or a zarr folder (it was not possible to allow both .tiff files and .zarr folders at the same time in a single dialog, as far as I could find). 
- I have added some explanatory text for the segmentation file/folder
- the area recompute should not pop up anymore if there is no segmentation.

I have additional updates on another branch, to allow importing of custom features for viewing in the plot. Maybe we want to merge this in here eventually as well. 

The example script for importing external tracks still needs to be updated, I will look into that tomorrow and also do some additional tests and checks.

cmalinmayor commented:
@AnniekStok My "final" reorganization is complete (I think). Some notes:

> points are now scaling automatically according to the x scale value. I also re-enabled setting it in the GUI, it seemed to be fine and I did not encounter the off-target effect any more (have to test more thoroughly still).
I meant scaling them into world units from pixel coordinates, not scaling them in the UI, sorry! I set the UI value back to the default of "5", and we can leave supporting import from pixel coordinates to the next PR.

Otherwise, I mostly moved all the widgets into their own files inside a menus folder, and then split up CSV and Segmentation related functionality into separate pages (the segmentation file browser is now merged with the scaling information page). 

I did try it with and without a segmentation on the hela example dataset, and also tried importing invalid data/options (selecting 3D for my 2D example). I'll approve this, and then when you get up, you can do a final test and merge if you like it!

cmalinmayor reviewed (approved):
🎉 😃
